controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -944,118 +944,158 @@ private boolean isSubscriberToNonSubscriberTransition(final ReaderGroupConfigRec
-        Map.Entry<StreamCutReferenceRecord, StreamCutReferenceRecord> limits =
-                getBoundStreamCuts(policy, retentionSet);
+        long currentTime = retentionClock.get().get();
+
+        Map.Entry<StreamCutReferenceRecord, StreamCutReferenceRecord> limits = getBoundStreamCuts(policy, retentionSet, 
+                x -> currentTime - x.getRecordingTime());
         // if subscriber lowerbound is ahead of streamcut corresponding to the max time and is behind stream cut for min time
         // from the retention set then we can safely truncate at lowerbound. Else we will truncate at the max time bound if it
         // exists
-        // 1. if LB > min => truncate at min
-        // 2. if LB < max => truncate at max
-        // 3. if LB < min && LB > max => truncate at LB
-        // 4. if LB < min && overlaps max => truncate at max
+        // 1. if LB is ahead of min => truncate at min
+        // 2. if LB is behind max => truncate at max
+        // 3. if LB is behind min && LB is ahead of max => truncate at LB
+        // 4. if LB is behind min && overlaps max => truncate at max
+        // 5. if LB overlaps with min and max ==> so its got both recent data and older data.. what do we do.. 
+        //      there are two options: either truncate beyond max or not truncate at all.  
         CompletableFuture<StreamCutRecord> limitMinFuture = limits.getValue() == null ? CompletableFuture.completedFuture(null) :
                 streamMetadataStore.getStreamCutRecord(scope, stream, limits.getValue(), context, executor);
 
         // if lowerbound is empty simply return min
         if (lowerBound == null || lowerBound.isEmpty()) {
             return limitMinFuture.thenApply(min -> Optional.ofNullable(min).map(StreamCutRecord::getStreamCut).orElse(null));
         }
+        Optional<StreamCutReferenceRecord> maxBoundRef = retentionSet
+                .getRetentionRecords().stream()
+                .filter(x -> currentTime - x.getRecordingTime() >= policy.getRetentionMax())
+                .max(Comparator.comparingLong(StreamCutReferenceRecord::getRecordingTime));
 
         CompletableFuture<StreamCutRecord> limitMaxFuture = limits.getKey() == null ? CompletableFuture.completedFuture(null) :
                 streamMetadataStore.getStreamCutRecord(scope, stream, limits.getKey(), context, executor);
-        return CompletableFuture.allOf(limitMaxFuture, limitMinFuture)
+        CompletableFuture<StreamCutRecord> maxBoundFuture = maxBoundRef.map(x -> streamMetadataStore.getStreamCutRecord(scope, stream, x, context, executor))
+                                                                    .orElse(CompletableFuture.completedFuture(null));
+        return CompletableFuture.allOf(limitMaxFuture, limitMinFuture, maxBoundFuture)
                 .thenCompose(v -> {
                     StreamCutRecord limitMax = limitMaxFuture.join();
```

Comment:
My understanding is that `limitMax` and `limitMin` are StreamCut Records that satisfy the min/max criteria for the 
 Stream. It may be good to explain that with a comment.